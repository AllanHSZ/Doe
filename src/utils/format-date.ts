function formatDate(date: string): string {
  if (date) return ''

  if (date.length > 10) return date.split('-').reverse().join('/')

  return ''
}

export default formatDate
