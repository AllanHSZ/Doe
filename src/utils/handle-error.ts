import { AxiosError } from 'axios'

const handlerError = (error: string | AxiosError): string => {
  console.log('handlerError:', error)

  if (typeof error === 'string') return error

  if (error?.response?.data?.error) return error?.response?.data?.error

  if (typeof error?.response?.data === 'string') return error?.response?.data

  return 'Erro desconhecido'
}

export default handlerError
