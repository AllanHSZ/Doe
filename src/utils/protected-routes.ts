import { GetServerSidePropsContext } from 'next'
import { DefaultSession, Session } from 'next-auth'
import { getSession } from 'next-auth/client'

export type MySession = {
  id: number
} & Pick<DefaultSession, 'expires' | 'user'>

async function protectedRoutes(
  context: GetServerSidePropsContext,
  session?: Session | null
) {
  if (!session) session = await getSession(context)

  if (!session?.user?.id) {
    context.res.writeHead(302, {
      Location: `/signin?callbackUrl=${context.resolvedUrl}`
    })
    context.res.end()
  }

  return session
}

export default protectedRoutes
