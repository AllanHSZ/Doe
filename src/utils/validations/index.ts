import Joi from 'joi'

const userFieldsValidations = {
  name: Joi.string().min(5).required().messages({
    'string.empty': 'Digite o seu nome.',
    'string.min': 'O nome precisa ter no mínimo 5 caracteres.'
  }),
  email: Joi.string()
    .email({ tlds: { allow: false } })
    .required()
    .messages({
      'string.empty': 'Digite o seu email.',
      'string.email': 'Email inválido.'
    }),
  password: Joi.string().min(6).required().messages({
    'string.min': 'A senha precisa ter no mínimo 6 caracteres.',
    'string.empty': 'Digite a sua senha.'
  }),
  confirm_password: Joi.string()
    .valid(Joi.ref('password'))
    .required()
    .messages({
      'any.required': 'Digite a senha novamente.',
      'any.only': 'As senhas não são iguais.'
    })
}

const addressFieldsValidations = {
  city: Joi.string().required().messages({
    'string.empty': 'Digite a cidade.'
  }),
  neighborhood: Joi.string().required().messages({
    'string.empty': 'Digite o bairro.'
  }),
  number: Joi.string().required().messages({
    'string.empty': 'Digite o número.'
  }),
  state: Joi.string().required().messages({
    'string.empty': 'Digite o estado.'
  }),
  street: Joi.string().required().messages({
    'string.empty': 'Digite a rua.'
  })
}

const donatinsFieldsValidations = {
  type: Joi.string().required().messages({
    'string.empty': 'Digite o tipo da doação.'
  }),
  description: Joi.string().required().messages({
    'string.empty': 'Digite a descrição da doação.'
  }),
  phone: Joi.string().required().messages({
    'string.empty': 'Digite um celular para contato.'
  }),
  email: Joi.string()
    .email({ tlds: { allow: false } })
    .required()
    .messages({
      'string.empty': 'Digite um email para contato.',
      'string.email': 'Email inválido.'
    })
}

export type FieldErrors = {
  [key: string]: string
}

function getFieldErrors(objError: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

type SignUpProps = {
  name?: string
  username?: string
  password?: string
  confirm_password?: string
}

export function signUpValidate(values: SignUpProps) {
  const { email: user, ...validations } = userFieldsValidations
  const schema = Joi.object({ user, ...validations })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type SignInProps = {
  username?: string
  password?: string
}

export function signInValidate(values: SignInProps) {
  const { email: username, password } = userFieldsValidations
  const schema = Joi.object({ username, password })

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type EditUserProps = {
  name?: string
  password?: string
  confirm_password?: string
}

export function editUserValidate(values: EditUserProps) {
  const { name, password, confirm_password } = userFieldsValidations

  const schema = Joi.object({
    name,
    ...(values?.password?.trim().length ? { password, confirm_password } : {})
  })

  return getFieldErrors(
    schema.validate(
      {
        name: values.name,
        ...(values.password
          ? {
              password: values.password,
              confirm_password: values.confirm_password
            }
          : {})
      },
      { abortEarly: false }
    )
  )
}

type AddressProps = {
  city: string
  neighborhood: string
  number: string
  state: string
  street: string
}

export function addressValidate(values: AddressProps) {
  const schema = Joi.object(addressFieldsValidations)

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}

type DonationValidationProps = {
  type: string
  description: string
  phone: string
  email: string
}

export function donationValidate(values: DonationValidationProps) {
  const schema = Joi.object(donatinsFieldsValidations)

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
