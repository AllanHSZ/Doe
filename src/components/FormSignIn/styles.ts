import styled, { css } from 'styled-components'
import { Form, FormErrorWrapper } from 'components/Form'

export const Wrapper = styled(Form)`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;

    ${FormErrorWrapper} {
      margin-top: ${theme.spacings.xxsmall};
    }
  `}
`

export const FieldsWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    width: 100%;

    & > div:last-of-type {
      margin-top: ${theme.spacings.xxsmall};
    }
  `}
`

export const ForgotPassword = styled.a`
  ${({ theme }) => css`
    margin-top: ${theme.spacings.xxsmall};
    font-size: ${theme.font.sizes.xsmall};
    color: ${theme.colors.textEmphasis};
    text-decoration: underline;
    text-align: end;
  `}
`
