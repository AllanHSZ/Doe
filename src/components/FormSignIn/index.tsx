import { useState } from 'react'
import { signIn } from 'next-auth/client'
import { useRouter } from 'next/router'
import { Lock, Mail } from 'react-feather'

import { useToast } from 'hooks/use-toast'
import { FieldErrors, signInValidate } from 'utils/validations'
import { FormError, FormLoading } from 'components/Form'
import Button from 'components/Button'
import Input from 'components/Input'

import * as S from './styles'

const FormSignIn = () => {
  const [formError, setFormError] = useState('')
  const [fieldError, setFieldError] = useState<FieldErrors>({})
  const [values, setValues] = useState({ username: '', password: '' })
  const [loading, setLoading] = useState(false)
  const routes = useRouter()
  const { push, query } = routes

  const showToast = useToast()

  if (query?.callbackUrl) {
    showToast({
      type: 'informative',
      message: 'Para acessar essa página é necessario estar conectado!'
    })
  }

  const handleInput = (field: string, value: string) => {
    setValues((s) => ({ ...s, [field]: value }))
  }

  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault()
    setFormError('')
    setLoading(true)

    const errors = signInValidate(values)

    if (Object.keys(errors).length) {
      setFieldError(errors)
      setLoading(false)
      return
    }

    setFieldError({})

    // sign in
    const result = await signIn('credentials', {
      ...values,
      redirect: false,
      callbackUrl: `${process.env.NEXT_PUBLIC_APP_URL}${
        query?.callbackUrl || ''
      }`
    })

    if (result?.error) {
      setLoading(false)
      setFormError(result.error)
      return
    }

    if (result?.url) {
      push(result?.url)
      return
    }

    setLoading(false)
    setFormError('Erro ao fazer login')
  }

  return (
    <S.Wrapper onSubmit={handleSubmit}>
      <Input
        name="username"
        type="email"
        label="Email"
        placeholder="Ex: joao.silva@email.com"
        error={fieldError?.user}
        onInputChange={(v) => handleInput('username', v)}
        disabled={loading}
        icon={<Mail />}
      />
      <Input
        name="password"
        label="Senha"
        placeholder="****"
        disabled={loading}
        error={fieldError?.password}
        onInputChange={(v) => handleInput('password', v)}
        type="password"
        icon={<Lock />}
      />

      <S.ForgotPassword href="#">Esqueceu a sua senha?</S.ForgotPassword>

      {!!formError && <FormError>{formError}</FormError>}

      <Button type="submit" fullWidth disabled={loading}>
        {loading ? <FormLoading /> : <span>ENTRAR</span>}
      </Button>
    </S.Wrapper>
  )
}

export default FormSignIn
