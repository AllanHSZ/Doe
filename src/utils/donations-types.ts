import { TextFieldOptions } from 'components/TextField'
// , ,
export default [
  {
    label: 'Acess√≥rio',
    value: 'ACESSORIO'
  },
  {
    label: 'Alimento',
    value: 'ALIMENTO'
  },
  {
    label: 'Roupa',
    value: 'ROUPA'
  }
] as TextFieldOptions[]
