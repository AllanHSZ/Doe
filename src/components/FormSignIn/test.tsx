import { render, screen } from 'utils/test-utils'

import FormSignIn from '.'

jest.mock('next/router', () => ({
  useRouter: jest.fn(() => ({ asPath: '/signin' }))
}))

describe('<FormSignIn />', () => {
  it('should render the form', () => {
    const { container } = render(<FormSignIn />)

    expect(screen.getByPlaceholderText(/email/i)).toBeInTheDocument()
    expect(screen.getByPlaceholderText(/\*\*\*\*/i)).toBeInTheDocument()

    expect(screen.getByRole('button', { name: /entrar/i })).toBeInTheDocument()

    expect(container.parentElement).toMatchSnapshot()
  })

  it('should render the forgot password link', () => {
    render(<FormSignIn />)

    expect(
      screen.getByRole('link', { name: /Esqueceu a sua senha\?/i })
    ).toBeInTheDocument()
  })
})
