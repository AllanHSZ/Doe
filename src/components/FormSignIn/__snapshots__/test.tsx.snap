// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSignIn /> should render the form 1`] = `
.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #F2F3F5;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #F2F3F5;
}

.c5:focus-within {
  box-shadow: 0 0 0.5rem #04D361;
}

.c7 {
  color: #394338;
  font-family: Lato,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.4rem;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c7:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #F2F3F5 inset;
  -webkit-filter: none;
  filter: none;
}

.c4 {
  font-size: 1.4rem;
  color: #394338;
  margin-bottom: 0.4rem;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 2.2rem;
  color: #8F8F8F;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c6 > svg {
  width: 100%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Lato,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border: 0;
  cursor: pointer;
  border-radius: 0.6rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #FAFAFA;
  background: #04D361;
  height: 4rem;
  font-size: 1.4rem;
  font-weight: 700;
  padding: 0.8rem 3.2rem;
  width: 100%;
}

.c10:hover {
  background: #04AB4F;
}

.c0 {
  width: 100%;
}

.c0 .c2 {
  margin: 1.2rem 0;
}

.c0 .c9 {
  margin-top: 1.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  margin-top: 0.8rem;
  font-size: 1.2rem;
  color: #212D22;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  text-align: end;
}

<body>
  <div>
    <form
      class="c0 c1"
    >
      <div
        class="c2 c3"
      >
        <label
          class="c4"
          for="username"
        >
          Email
        </label>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
              />
              <polyline
                points="22,6 12,13 2,6"
              />
            </svg>
          </div>
          <input
            class="c7"
            id="username"
            name="username"
            placeholder="Ex: joao.silva@email.com"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        class="c2 c3"
      >
        <label
          class="c4"
          for="password"
        >
          Senha
        </label>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="11"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="11"
              />
              <path
                d="M7 11V7a5 5 0 0 1 10 0v4"
              />
            </svg>
          </div>
          <input
            class="c7"
            id="password"
            name="password"
            placeholder="****"
            type="password"
            value=""
          />
        </div>
      </div>
      <a
        class="c8"
        href="#"
      >
        Esqueceu a sua senha?
      </a>
      <button
        class="c9 c10"
        type="submit"
      >
        <span>
          <span>
            ENTRAR
          </span>
        </span>
      </button>
    </form>
  </div>
</body>
`
